pages/blog/the-reference-rabbit-hole.md

```diff
@@ -0,0 +1,275 @@
+
+Well, that was easy, I see no rabbit hole here Jonas!? 
+
+# Welcome to the rabbit hole
+During the discussion, Sergio brought up that Fran was actually using an illegal reference, as he, in one of the examples, was using a **Reference Object** for a server, which was not allowed.  More specifically, it was this example where he references the `mosquitto` server:
+
+```yaml
+...
+servers:
+  mosquitto:
+    $ref: 'common.asyncapi.yaml#/components/servers/mosquitto'
+```
+
+My immediate reaction was "wait .... It's not?!". 
+
+I had always used `$ref` quite extensively in my AsyncAPI documents and specifically used a reference for servers. And I knew that the tooling had no problems with the `$ref` as long as it was a valid reference. 
+
+But Sergio was absolutely right, a second look into the specification, `servers` are defined using [Servers Object](https://www.asyncapi.com/docs/specifications/v2.2.0#serversObject), which are defined using a map of [Server Object](https://www.asyncapi.com/docs/specifications/v2.2.0#serverObject)s. **NOT** `Server Object | Reference Object` as I expected.
+
+So after that, we started to realize, that there is quite a big difference when and where **Reference Object**s are allowed. For the full list of discrepancies, check out [spec #650](https://github.com/asyncapi/spec/issues/650).
+
+But... Why did I think it was allowed to do so? 
+
+## Discrepancies in AsyncAPI Tooling
+So back to my own experience, why was I so sure that the tooling allowed for me to use **Reference Object**s for servers? 
+
+Well, as it turns out, it is because the [JS parser](https://github.com/asyncapi/parser-js) dereferences before it validates the AsyncAPI document. This means that if I defined my AsyncAPI document such as:
+
+```yaml
+asyncapi: '2.2.0'
+...
+servers:
+  test-server:
+    $ref: './servers/testServer.yaml'
+...
+```
+
+Together with `testServer.yaml`:
+```yaml
+url: ws://mycompany.com/ws
+protocol: ws
+```
+
+Validating the AsyncAPI document using a tool such as [ajv](https://ajv.js.org/) against [JSON Schema representation for 2.2.0](https://github.com/asyncapi/spec-json-schemas/blob/master/schemas/2.2.0.json), it would reject it. 
+
+However, because the parser dereferences first, the document that is being validated is this:
+
+```yaml
+asyncapi: '2.2.0'
+...
+servers:
+  test-server:
+    url: ws://mycompany.com/ws
+    protocol: ws
+...
+```
+
+Checkout [parser-js #405](https://github.com/asyncapi/parser-js/issues/405) for more information.
+
+## What about `$id` keyword
+
+The next thoughts that came to out minds was, one of the key differences between our **Reference Object**, and how `$ref` is resolved in JSON Schema Draft 7, is the [$id keyword](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-01#section-8.2). This allows you to define a URI that is used as a base URI. This means that for example a message such as this:
```

Comment:
I need to suggest you to remove the introductory part of this sentence. It is so long and couldn't follow it without reading 3 times.
```suggestion
One of the key differences between our **Reference Object**, and how `$ref` is resolved in JSON Schema Draft 7, is the [$id keyword](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-01#section-8.2). This allows you to define a URI that is used as a base URI. This means that for example a message such as this:
```

Reply:
Makes it much more clear 👍